### PR TITLE
feat: Send image URLs to CDN.

### LINKED ISSUE #167: Tile Images to CDN (Contile image hosting)
Store images provided in the adM Instant Suggest tiles to our own Content Distribution Network (CDN) system. This is the top priority for Contile v2.

Business need: Updates to existing built-in advertiser icons have to ride the train, which is too long of a delay if an icon needs to be updated for any reason. The ability to update images off-train is also a key value prop of Contile.

Privacy need: This would increase user privacy by potentially not leaking information to an unknown CDN.

[https://github.com/mozilla-services/contile/issues/19](https://github.com/mozilla-services/contile/issues/19)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-33)
┆Epic: Contile v2

### PR DESCRIPTION
**Pull request by jrconlin**
Closes #167

## Description

Send ADM provided images to a CDN.

This will activate the `img_storage` mod, and the `storage` JSON config option. 

## Testing

TODO: Add unit tests for 
* config options
* CDN image load
* validate that img URLs are getting swapped appropriately

## Issue(s)

Closes #167

### CODE REVIEW

File: src/server/img_storage.rs
Comment (ncloudioj) [inline src/server/img_storage.rs:253]: For context, per newtab specs, images for TopSites should be square, if `heigh != width`, we can treat it as an invalid image. Also, the minimal hi-res icon size for TopSites is 96 pixels, so if it's below that, we can also treat it as invalid.

Assuming we're going to use `image_url` for the URL, shall we call it `image_size` for this one?

Comment (jrconlin) [reply src/server/img_storage.rs:253]: Works for me! Thanks!

are there any other things we should check the image for? (e.g. max size?, unusual format? (not "jpg" or "png") etc.)

Comment (ncloudioj) [reply src/server/img_storage.rs:253]: I think we can enforce the following things here:
* Size: `width == height && 96 <= wide <= 256`
* MIME extensions: Firefox only takes any of `png, jpg, svg`

**Comment by ncloudioj** [comment]
@jrconlin Thanks for working on this feature!

Just wanted to add a quick note that when we upload the image to GCS, let's make sure the following headers are attached as well

```
{
    'Cache-Control': 'public, max-age=31536000', // cache for 1 year
    'Content-Disposition': 'inline' // show this image inline on the page other than showing a "Save as" prompt
}
```

**Comment by jrconlin** [comment]
I know that the CDN uses the `Cache-Control` header we set for the bucket, but not sure how to set the `Content-Disposition` header.

**Comment by ncloudioj** [comment]
Looks like `Content-Disposition` is supported by its [XML API](https://cloud.google.com/storage/docs/xml-api/reference-headers). IIRC, `inline` is the default value for `Content-Disposition`, if our library doesn't support it, let's just keep an eye on the browser making sure it behaves properly.

File: src/adm/tiles.rs
Comment (pjenvey) [inline src/adm/tiles.rs:106]: Is this field still needed when `ImageMetrics` has a size?

We _could_ also cleanup this response output but it's not terribly important (nor do I know how it would affect the client impl cc: @ncloudioj)

e.g. 
```json
  "click_url": "...",
  image: {
    "url": "...",
    "size": "...",
    "width": "etc",
  }
```

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:122]: Might wanna also kill these unused fields (like `hash`) until they're needed.

Another general nit -- I've been trying not to include "Result" in struct names such as this one as it makes them sound like they're a variant of a `Result` (such as `HandlerResult`) to fresh eyes. Maybe `StoredImage`

**Review by pjenvey — CHANGES_REQUESTED** [review]

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:68]: nit: since StorageSettings has a `Default` impl we shouldn't need these functions and their associated `#[serde(default = "default_ttl")]`.

Instead we can use the `#[serde(default)]` container attribute on the entire struct (see above) to fill these in from the `Default` impl.

File: src/adm/tiles.rs
Comment (jrconlin) [reply src/adm/tiles.rs:106]: Part of the confusion is that "size" apparently is different. The UA wants "size" to be the width/height. We want "size" to be the length of the image (to construct the file name).

Comment (ncloudioj) [reply src/adm/tiles.rs:104]: Firefox only expects the `image_url` and the `image_width` from Contile. By historical reason, we reference `image_width` as `image_size` in the Fx codebase.

Alternatively, We can also leave out `image_width` from the response payload if Contile can enforce this invariant:
Contile would only provide the `image_url` field in the response if `image.width == image.height &&  96 <= image.width <= 256`.

So when we see `image_url` in the response, we know that the image is of the valid format, we can simply assign the minimal favicon width to it, and Firefox knows how to render it properly.

@jrconlin Let me know which one you prefer so that I can adjust my client patch accordingly.

Comment (jrconlin) [reply src/adm/tiles.rs:104]: I'll add this as a comment to the code and only return `image_url` and `image_size`.

Comment (ncloudioj) [reply src/adm/tiles.rs:104]: sounds good!

Comment (pjenvey) [reply src/adm/tiles.rs:108]: > Alternatively, We can also leave out `image_width` from the response payload if Contile can enforce this invariant:
> Contile would only provide the `image_url` field in the response if `image.width == image.height && 96 <= image.width <= 256`.

@jrconlin It looks like we're doing exactly this, so it sounds like we could only return `image_url`?

Comment (jrconlin) [reply src/adm/tiles.rs:108]: I'd like to return both, since the size could vary between 96 and 128 px.
(I'll also say that previous experience has taught me to return more info than I need in these cases, but I'll defer to the folk doing the work.)

Comment (ncloudioj) [reply src/adm/tiles.rs:108]: Agreed, let's also return `image_size` other than letting UA hardcode it with `MIN_FAVICON_SIZE (96)`.

BTW, Dao and Amy have reviewed the client-side [patch](https://phabricator.services.mozilla.com/D119923), I am landing it to nightly now.

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:267]: - Let's give `StorageImage` a reference to the `reqwest_client` built at `Server` creation and call reqwest through that instead (to take advantage of the underlying per host pooling)
- Probably worth adding a timeout to the get call like we do for adM
- Definitely want an `.error_for_status()?` for the get too
- Reqwest errors have an `Into<HandlerError>` so map_err shouldn't be necessary (the reqwest error and its traceback should tell us enough about this error)

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:401]: Nothing to hold this up over but I think some of the Internal and BadImage errors in here derived from `cloud_storage::Error` should have their own `HandlerKind` `CloudStorage(#[from] cloud_storage::Error)` variant

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:500]: debugging nit:

This assert ends up hiding the error, this'll show it:

```suggestion
        bucket.store(&target).await.expect("Store failed");
```

File: src/server/img_storage.rs
Comment (pjenvey) [inline src/server/img_storage.rs:366]: :smart:

**Review by pjenvey — CHANGES_REQUESTED** [review]
I'm still looking at the filename hashing bits but this is looking good

File: src/server/img_storage.rs
Comment (jrconlin) [reply src/server/img_storage.rs:401]: Yeah, I figure we can revisit and revise the errors later, probably when we break things up to componentize stuff.  For now, having the one, semi-universal errors is "good enough".

**Review by pjenvey — APPROVED** [review]

**Comment by pjenvey** [comment]
r+ sorry for the delay

**Pull request merged**
